existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,292 @@
+
+  -- log to a logfile
+  func (appendFile "logfile.log" . show)
+```
+We will see some nasty compiler errors:
+```haskell
+  log username
+      ^^^^^^^^
+Couldn't match expected type `a` with actual type `String`...
+```
+
+We're getting this error because:
+1. `func`'s type variable `a` is universally quantified, therefore it can only be instantiated at the use-site.
+2. We're trying to instantiate it at `func`'s definition-site (`log username` is trying to instantiate the type variable `a` with `String` and `log userCount` is trying to instantiate the type variable `b` with `Int`).
+Because we want `a` to be instantiated at `func`'s definition-site (rather than its use-site), we need `a` to be existentially quantified instead.
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
+With this change we can instantiate `a` when we call log.
+
+```haskell
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+func log = do
+  username <- getUsername
+  log @String username
+
+  userCount <- getUserCount
+  log @Int userCount
+```
+
+Note, that this type applications is unnecessary.
+Compiler is smart enough to choose appropriate types.
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
+Let's take a look how this is done in datatypes.
+You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.
+
+```haskell
+data Elem = forall a. Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+```
+
+This is a heterogeneous list that contains values of different types.
+You can't do that with an universally quantified list because every element in that list would need to have the same type.
+But, there are not many things you can do with elements of this list.
+Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.
+However, adding constraints to the `a` variable may improve the situation:
+
+```haskell
+-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
+data Elem = forall a. (Show a) => Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+
+-- We can use 'print' here because inner types of 'Elem' has 'Show' constraint
+printElem :: Elem -> IO ()
+printElem (Elem x) = print x
+
+main :: IO ()
+main = fmap printElem multitypedList
+
+λ> "a"
+λ> 1
+λ> Just 5
+```
+Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.
```

Comment:
This statement and the example below feels a bit weird. Isn't it obvious that you cannot add constraints to a type variable that is not in scope? It is also not clear why we are trying to do that here.

Reply:
Well, it is, but why not to mention it? I mean, It is better to explicitly state that it is impossible, for those who still hope, that there is a way...

Reply:
>but why not to mention it?

For sake of brevity I guess. Basically in articles like these, it does not often help to put as much detail as possible, without carefully arranging it. So unless there is a really good reason to think that the reader will naturally wonder if they can add a constraint to a hidden type variable, I think it is better to omit it. Not a big deal though, I ll leave it upto you.